### PR TITLE
214issue : Fix sidebar for improved navigation

### PR DESCRIPTION
**Pull request by Vivek-04022001**
### **Here are the changes I've made:**

_1.Within the 'sidebar' component, I've implemented modifications that fix the LanguagePicker and TagPicker components in place.
2. I've introduced a new component named ScrollToTopButton, responsible for managing both the user interface and functionality of the button that scrolls the page to the top.
3. Inside the Layout component, I've integrated the ScrollToTopButton.
 Here is the demo of the feature._

https://github.com/lucavallin/first-issue/assets/91373023/b3478a71-c414-4246-8fe0-07244bdec4d9

### CODE REVIEW

File: components/Layout.tsx [hunk 1]
```diff
@@ -1,19 +1,20 @@
...
-
 type LayoutProps = {
   children: React.ReactNode;
 };
 
 export const Layout = ({ children }: LayoutProps) => (
   <div className="flex min-h-screen flex-col bg-dark-400 text-light-300 antialiased">
     <Header />
-    <main className="flex flex-1">
+    <main className="flex flex-1 ">
```
Comment (lucavallin) [inline components/Layout.tsx:13]:
```suggestion
    <main className="flex flex-1">
```

File: components/ScrollToTopButton.tsx [hunk 2]
```diff
@@ -0,0 +1,46 @@
+import { faArrowUp } from "@fortawesome/free-solid-svg-icons";
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+import { useEffect, useState } from "react";
+const ScrollToTopButton = () => {
+  const [showUpArrow, setShowUpArrow] = useState(false);
+
+  // Function to handle the scroll event
+  const handleScroll = () => {
+    // Check the scroll position to determine when to show the "Up Arrow" button
+    if (window.scrollY > 700) {
```
Comment (lucavallin) [inline components/ScrollToTopButton.tsx:10]: Please add comment on why `700`

File: components/ScrollToTopButton.tsx [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+  // Function to scroll to the top of the page
+  const scrollToTop = () => {
+    window.scrollTo({ top: 0, behavior: "smooth" });
+  };
+
+  return (
+    <div>
+      {showUpArrow && (
+        <button
```
Comment (lucavallin) [inline components/ScrollToTopButton.tsx:35]: We could also just hide it by applying the `hide` class as needed, might be cleaner.

File: components/Sidebar.tsx [hunk 4]
```diff
@@ -15,17 +15,36 @@ export const Sidebar = () => {
   const { languages, tags } = useAppData();
   const { tag: activeTagId, language: activeLanguageId } = router.query;
   const pageName = router.pathname.split("/")[1];
+  const [scrollHeightReached, setScrollHeightReached] = useState(false);
 
   useEffect(() => {
     const scrollTarget = document.getElementById("repositories-list");
     const isMobile = window.innerWidth <= 640;
```
Comment (lucavallin) [inline components/Sidebar.tsx:22]: Please document all these hardcoded numbers

File: components/Sidebar.tsx [hunk 5]
```diff
@@ -15,17 +15,36 @@ export const Sidebar = () => {
...
     if (isMobile && (pageName === "language" || pageName === "tag")) {
```
Comment (lucavallin) [inline components/Sidebar.tsx:23]: Please add comment explaining what is happening here.

File: components/Sidebar.tsx [hunk 6]
```diff
@@ -15,17 +15,36 @@ export const Sidebar = () => {
...
       scrollTarget?.scrollIntoView({ behavior: "smooth" });
     }
+
+    // Function to handle the scroll event
+    const shouldApplyStickyClasses = () => {
+      // Check the scroll position to determine when to apply classes
+      if (window.scrollY >= 701.5) {
```
Comment (lucavallin) [inline components/Sidebar.tsx:30]: Please document all these hardcoded numbers

File: components/Sidebar.tsx [hunk 7]
```diff
@@ -15,17 +15,36 @@ export const Sidebar = () => {
...
+        setScrollHeightReached(true);
+      } else {
+        setScrollHeightReached(false);
+      }
+    };
+
+    // Add the scroll event listener
+    window.addEventListener("scroll", shouldApplyStickyClasses);
```
Comment (lucavallin) [inline components/Sidebar.tsx:38]: We have similar code in the scroll component itself, can they be unified somehow?

File: components/Sidebar.tsx [hunk 8]
```diff
@@ -15,17 +15,36 @@ export const Sidebar = () => {
...
+
+    // Remove the scroll event listener when the component unmounts
+    return () => {
+      window.removeEventListener("scroll", shouldApplyStickyClasses);
+    };
   }, [router.asPath, pageName]);
 
   return (
-    <section className="w-full flex-none px-6 font-sans text-light-300 md:max-w-sm">
+    <section className=" w-full flex-none px-6 font-sans text-light-300 md:relative md:max-w-sm">
```
Comment (lucavallin) [inline components/Sidebar.tsx:47]:
```suggestion
    <section className="w-full flex-none px-6 font-sans text-light-300 md:relative md:max-w-sm">
```

**Review by lucavallin — COMMENTED** [review]
UI looks great. Left a few comments.

File: components/ScrollToTopButton.tsx [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+import { useEffect, useState } from "react";
+const ScrollToTopButton = () => {
+  const [showUpArrow, setShowUpArrow] = useState(false);
+
+  // Function to handle the scroll event
+  const handleScroll = () => {
+    // Check the scroll position to determine when to show the "Up Arrow" button
+    if (window.scrollY > 700) {
+      setShowUpArrow(true);
```
[the remark this answers is not in the document]
Comment (Vivek-04022001) [reply components/ScrollToTopButton.tsx:11]: I further optimize the code : 
'''
 // Function to handle the scroll event
  const handleScroll = () => {
    
    // Check the scroll position to determine when to show the "Up Arrow" button. 
    // The "ScrollToTopButton" is only visible when the user scrolls to a minimum of 700px vertically.
    setShowUpArrow(window.scrollY > 700);
    
  };
'''

**Comment by Vivek-04022001** [comment]
### ** 🫡@lucavallin, I've completed all the review changes. Please take a look when you have a moment. Thank you!**

**Comment by lucavallin** [comment]
Uploading Screen Recording 2023-10-11 at 11.46.48.mov…

@Vivek-04022001 Nice, but it seems the scrolling is broken.

**Comment by Vivek-04022001** [comment]
@lucavallin , I'm sorry. I will try to solve this issue

**Comment by lucavallin** [comment]
image

build error

**Comment by Vivek-04022001** [comment]
### **Fix : prop types for ScrollToTopButton.**
```
// Define a type for the props that the ScrollToTopButton component accepts.
type ScrollToTopButtonProps = {
👉  handleOnClick: () => void;
};

const ScrollToTopButton = ({ handleOnClick }: ScrollToTopButtonProps) => {
  return (
    <div>
      <button
        onClick={handleOnClick}
        className="fixed bottom-4 right-4 flex h-14 w-14 items-center justify-center rounded-full border border-primary bg-transparent hover:bg-primary md:bottom-8 md:right-10"
      >
        <FontAwesomeIcon icon={faArrowUp} size="xl" />
      </button>
    </div>
  );
};
```

**Comment by Vivek-04022001** [comment]
### @lucavallin, I kindly request your review of the recent updates I've made.

**Review by lucavallin — APPROVED** [review]

**Pull request merged**